First draft of the juttle tutorial with github data

dmehra opened the pull request:
@demmer this is still raw, want high level review. Missing:
- join section that enriches the data with metadata
- at some point, switch to use elastic for the data with big data set (ingest with `read file | write elastic` ?)
- haven't validated all the juttles that run in outrigger

Ignore exact wording, not ready for review of that level yet. Mainly want review of "does this tutorial cover the right bits of juttle". Most of it came from the old tutorial, with some mods.

dmehra commented:
How about review by people who're here this week: @rlgomes @go-oleg @davidvgalbraith

rlgomes commented on docs/concepts/juttle_tutorial.md:
```diff
@@ -2,27 +2,532 @@
+
+First, let's look at a few events using the `read file` source. We already used `head` and `tail` processors to limit the data we output, now we will try filtering by timestamp. Juttle defaults to looking for the timestamp in the field `time` unless `-timeField <fieldname>` option is used.
+
+XXX/dmehra this is currently unsupported:
```
@dmehra I can add this feature to the file adapter but would want to wait for this other PR to get in before fiddling in that are of the code: https://github.com/juttle/juttle/pull/11

rlgomes commented on docs/concepts/juttle_tutorial.md:
````diff
@@ -2,27 +2,532 @@
+There is a Juttle trick to see which fields are present in the data; the processing will not be optimized, so do this only over small data sets (by specifying a narrow time range, for example).
+
+```
+read file -file '/tmp/data.json' -from :2015-07-01: -to :2015-07-02:
````
seems like we REALLY need the `-from/-to` handling to get in so we can use it here in the tutorial.

dmehra replied:
Yes thanks for adding that feature, I'll pull master into this branch and redo the tutorial example with working from/to.

go-oleg commented:
Overall I think this does a great job of introducing juttle.

Some thoughts
 - Where will a reader typically be coming from when reading this tutorial? I think having a basic understanding of the anatomy of a juttle program [shown here](http://juttle.github.io/juttle/concepts/overview/) is prereq for going through this tutorial.
 - The to-be-added `join` example will also serve to demonstrate that a flowgraph can start with two sources. Perhaps it would be good to have an example in here that shows a flowgraph can end in two sinks as well (all the current examples branch and merge in the middle).
 - Weave more links to "advanced" features of juttle (e.g. features like functions, subs, and user defined reducers which allow for more modular programming).

davidvgalbraith commented on docs/concepts/juttle_tutorial.md:
```diff
@@ -2,27 +2,532 @@
+
+When working with a new dataset, one commonly wants to know how big the data is, and how it's structured in terms of fields and values. We know the total number of points in our data set, and we have an idea of what fields are in the points from doing `read ... | head` to look at a few points. However, those points may not have represented full variety of the data set; there may be other points i ...
+
+There is a Juttle trick to see which fields are present in the data; the processing will not be optimized, so do this only over small data sets (by specifying a narrow time range, for example).
```
Might be handy to define `small data sets` here.

dmehra replied:
Could use help with the definition: would it be "data set that can fit in memory of a Node.js process"? what's the cap - 1GB? and how much of that can be used up by the data?

rlgomes replied:
we could simply link the following:
https://github.com/nodejs/node-v0.x-archive/wiki/FAQ#what-is-the-memory-limit-on-a-node-process

dmehra replied:
:white_check_mark:

davidvgalbraith commented on docs/concepts/juttle_tutorial.md:
```diff
@@ -2,27 +2,532 @@
+
+Let's say we wanted to know the number of actors participating in an average GitHub repository. The program above gives us the counts of actors and repos, so finding the average is simple math. However, it will not work to add "avg_actors = actors / repos" to the `reduce` expression, since this computation is not a reducer; and in general, this is not the logic you seek.
+
+If your intention is to add the computed field to each data point, thereby enriching and widening the dataset rather than reducing it, then it's not `reduce` you should be using but [`put`](../processors/put.md) processor.
```
Should be "it's not reduce you should be using but THE put processor"

dmehra replied:
Will fix.

davidvgalbraith commented on docs/concepts/juttle_tutorial.md:
```diff
@@ -2,27 +2,532 @@
+
+## Enriching the data (Put vs reduce)
+
+Let's say we wanted to know the number of actors participating in an average GitHub repository. The program above gives us the counts of actors and repos, so finding the average is simple math. However, it will not work to add "avg_actors = actors / repos" to the `reduce` expression, since this computation is not a reducer; and in general, this is not the logic you seek.
```
This is a bit opaque, what does it mean for a computation to be a reducer?

dmehra replied:
I had trouble with this bit, there's a conversation going on gitter, https://gitter.im/juttle/juttle?at=5682d4180199d70069e01f47
My best idea right now is to simply link from this to the definition of our reducers.

davidvgalbraith commented on docs/concepts/juttle_tutorial.md:
````diff
@@ -2,27 +2,532 @@
+| reduce actors = count_unique('actor_login') by repo_name
+```
+
+If you want to see the top 10 repositories by number of participating actors, append the [`sort`](../processors/sort.md) processor with -desc order, then limit to first 10 points with the [`head`](../processors/head.md) processor. 
````
We lapsed from first to second person here, let's make this just start with "To see the top 10 ...".

dmehra replied:
Will fix.

davidvgalbraith commented on docs/concepts/juttle_tutorial.md:
````diff
@@ -2,27 +2,532 @@
+| put activity_per_actor = Math.round(events/actors) by repo_name
+```
+
+This example clearly shows that `put` adds the newly computed field to each data point. Also, here we are using `put` with by-grouping (to be fair, the program above would run identically even with plain `put`, but it helps express the intention of the code).
````
This parenthetical throws me off, let's replace it with a few words about what by-grouping does in general without admitting that it's not doing anything in this program.

dmehra replied:
Will fix.

go-oleg commented on docs/concepts/juttle_tutorial.md:
```diff
@@ -2,27 +2,532 @@
+read file -file '/tmp/data.json'
+| filter sub_category = 'Microservice Platform'
+| reduce -from :2015-01-01: -over :w: -every :d: count() by 'repo_name'
+| @timechart -keyField 'repo_name'
```
s/@timechart/view timechart/

dmehra replied:
:white_check_mark: in all places.

go-oleg commented on docs/concepts/juttle_tutorial.md:
```diff
@@ -2,27 +2,532 @@
+read file -file '/tmp/data.json'
+| filter sub_category = subcat
+| reduce -from :2015-01-01: -over :w: -every :d: count() by 'repo_name'
+| @timechart -keyField 'repo_name'
```
s/@timechart/view timechart/

dmehra replied:
:white_check_mark:

dmehra commented:
Addressed the review comments (thanks guys) except these two:

- [ ] add example that shows a flowgraph can end in two sinks - will think about where to fit it
- [ ] Weave more links to "advanced" features of juttle (e.g. features like functions, subs, and user defined reducers which allow for more modular programming). - unclear on this one:

@go-oleg did you mean to just add them as "juttle has these features but tutorial doesn't cover them", or to actually include them in the tutorial? If the latter, i'm not coming up with useful places to put a function here, much less a reducer. There are places where I can utilize a sub.

Sections on join and ES are new and need review.

@henridf your review would be most welcome, as you fathered the first version of this tutorial (a lot of it stayed true to the original).

go-oleg commented:
I was hoping throughout the tutorial, but you're right in that its going to be complicated. Maybe some links at the end of the tutorial along the lines of "go here next":
 - list of built in procs
 - subs and functions
 - custom reducers

henridf commented:
A thought for further out - once there is a Google BigQuery adaptor, it would be compelling to use that to pull data rather than use a static file.

dmehra commented:
TODO:
- [ ] change ES data to use single-index schema instead of daily
- [ ] fix up the ending of tutorial with inputs as soon as they work in outrigger

dmajda commented:
As the tutorial work moved to Outrigger, I suppose this PR won’t ever get merged ⇒ closing.

@dmehra If I am mistaken, please reopen.

dmehra commented:
Will reopen because we still want the main tutorial doc merged into juttle for publishing as part of the docs site, I think, so all local docs links would work (and a big part of this doesn't use outrigger).
